src/models.py

```diff
@@ -16,24 +24,69 @@
 
 logger = logging.getLogger(__name__)
 
+class MLP(nn.Module):
+    def __init__(self, num_hidden):
+        super().__init__()
+        self.num_hidden = num_hidden
+
+        self.lin1 = nn.Linear(num_hidden, num_hidden)
+        self.relu1 = nn.ReLU(inplace=True)
+       
+    def forward(self, x):
+        x = self.lin1(x)
+        x = self.relu1(x)
+        return x
+
+class ResNet(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.intermediate = {}  # Activation(s) of intermediate layers
+        # Pre-trainer Resnet34
+        self.pretrained = models.resnet34(pretrained=True)
+        # Attaching a hook to the last layer with informative feature vector
+        self.pretrained.layer4[2].bn2.register_forward_hook(self.forward_hook('BatchNorm'))
+        
+    def forward_hook(self,layer_name):
+        def hook(module, input, output):
+            self.intermediate.update({layer_name: output})
+        return hook
+
+    def forward(self, input):
+        out = self.pretrained(input)
+        return out, self.intermediate
+
+
 class EncoderDecoder(nn.Module):
     def __init__(self, encoder, decoder):
         super().__init__()
         self.encoder = encoder
+        self.mlp1 = MLP(decoder.num_hidden)
+        self.mlp2 = MLP(decoder.num_hidden)
         self.decoder = decoder
         
     def forward(self, img, caption, seq_lengths, targets=None):
-        img_features = self.encoder(img)  # shape: (batch, img_features)
-        img_features = img_features.repeat(2, 1, 1)  # shape: (num_layers, batch, img_features)
-#         import pdb; pdb.set_trace()
-        logits = self.decoder(caption, img_features, seq_lengths) 
-        
+        img_features, intermediate_layer1 = self.encoder(img)  
```

Comment:
This connects to my comment of line 56 -- you could acces the intermediate layer immediately:
```suggestion
       img_features = self.encoder(img)
       intermediate_layer1 = self.encoder.intermediate
```

``

Reply:
ok